### PR TITLE
Add a profile endpoint

### PR DESCRIPTION
**Pull request by iliadsh**

### CODE REVIEW

**Comment by iliadsh** [comment]
For the set behavior, I needed to manually check/create each node in the path with separate queries to ensure the whole path exists. I'm wondering if it's possible to do in a single query. I see that in Postgres 14 there is a jsonb [subscript](https://www.postgresql.org/docs/14/datatype-json.html#JSONB-SUBSCRIPTING) operator which will attempt to create the whole path, but even then I'm unsure how to do that in a prepared statement.

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    this.servletRequest = servletRequest;
+    this.profileService = profileService;
+    this.userFactory = userFactory;
+  }
+
+  private SamUser getUser() {
+    return userFactory.from(servletRequest);
+  }
+
+  private List<String> parsePath(String path) {
```
Comment (melissachang) [inline service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: We have some conventions

- Public methods before private methods
- If method A calls helper method B, B is after A

So move getUser, parsePath to bottom

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  @Override
+  public ResponseEntity<Void> userProfileSet(AnyObject body, String path) {
+    profileService.setProperty(getUser(), parsePath(path), body.getValue());
+
+    return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+  }
+
+  @Override
+  public ResponseEntity<AnyObject> userProfileLookup(String path) {
+    var obj = profileService.getProperty(getUser(), parsePath(path));
```
Comment (melissachang) [inline service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: We have a convention to not use var if type isn't obvious, like here and below

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  private List<String> parsePath(String path) {
+    if (path == null || path.length() == 0) return List.of();
+
+    return Arrays.asList(path.split("\\."));
+  }
+
+  @Override
+  public ResponseEntity<Void> userProfileSet(AnyObject body, String path) {
```
Comment (melissachang) [inline service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: nit WSM has [`createWorkspace`](https://workspace.dsde-dev.broadinstitute.org/swagger-ui.html#/Workspace/createWorkspace), `getWorkspace`, etc. Can we rename to `setUserProfile`, `getUserProfile`.

File: service/src/main/java/bio/terra/user/db/ProfileDao.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    try {
+      return jdbcTemplate.queryForObject(
+          sql,
+          params,
+          (rs, rowNum) -> {
+            return rs.getString("value");
+          });
+    } catch (EmptyResultDataAccessException e) {
+      throw new PropertyNotFoundException("", e);
```
Comment (melissachang) [inline service/src/main/java/bio/terra/user/db/ProfileDao.java]: It's nice to have a helpful error message

"Property %s not found for user %s"

File: service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  @Autowired private MockMvc mockMvc;
+
+  @MockBean private ProfileSamUserFactory userFactory;
+
+  private final SamUser user = mock(SamUser.class);
+
+  @BeforeEach
+  void beforeEach() {
+    when(userFactory.from(any())).thenReturn(user);
+    var id = String.valueOf(Math.random()).substring(2);
```
Comment (melissachang) [inline service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java]: We usually have a helper method to generate random numbers, that way you don't have to type this every time

https://cs.github.com/DataBiosphere/terra-workspace-manager/blob/792780ab6d1bf38ee9e4b811811952135974c7f2/integration/src/main/java/scripts/testscripts/FolderLifecycle.java?q=appendrandomnumber#L74

Though in this case, does subject have to be different every time? Can you do

```
when(user.getSubjectId()).thenReturn("fake-12345");
```

File: service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  @Test
+  void getEmptyProfile() throws Exception {
+    mockMvc.perform(get(API)).andExpect(status().isNotFound());
+  }
+
+  @Test
+  void recursiveSet() throws Exception {
+    String json = "{ \"value\": \"John\" }";
```
Comment (melissachang) [inline service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java]: It makes these tests easier to read if you write helper methods for the mockMvc calls ([example](https://cs.github.com/DataBiosphere/terra-workspace-manager/blob/792780ab6d1bf38ee9e4b811811952135974c7f2/service/src/test/java/bio/terra/workspace/app/controller/WorkspaceApiControllerConnectedTest.java#L165))

```
  @Test
  void recursiveSet() throws Exception {
    setUserProfile("user.name.first", "John");
    assertUserProfile("user.name.first", "John");
  }
```

Also, consider combining tests if there's overlapping setup.

```
  @Test
  void setUserProfile() throws Exception {
    setUserProfile("user.name.first", "John");
    assertUserProfile("user.name.first", "John");

    // Existing property is updated
    setUserProfile("user.name.first", "Mary");
    assertUserProfile("user.name.first", "Mary");
  }
```

File: service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        .perform(get(API))
+        .andExpect(status().isOk())
+        .andExpect(content().contentType(MediaType.APPLICATION_JSON))
+        .andExpect(jsonPath("$.value.settings.darkmode").value(true))
+        .andExpect(jsonPath("$.value.settings.language").value("en"));
+  }
+
+  @Test
+  void overwrite() throws Exception {
+    String json = "{ \"value\": {\"a\": true, \"b\": false } }";
```
Comment (melissachang) [inline service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java]: If I understand this correctly, how about some more assertions

```
@Test
void writeTree() throws Exception {
  String tree = "{\"a\": true, \"subtree\": {b: false} }";
  setUserProperty( /*path=*/, tree);
  assertUserProperty(); // Assert entire tree
  assertUserProperty(); // Assert just a
  assertUserProperty(); // Assert just b inside subtree
}
```

**Review by melissachang — COMMENTED** [review]
Nice!

nit - We typically put Jira ticket # in PR title, [eg here](https://github.com/DataBiosphere/terra-workspace-manager/pull/854). This way, people can look at Jira issue for more context. Also, Jira ticket has links to PRs.

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java (hunk 2 above)
Comment (cbookg) [reply service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: FWIW I really dislike this convention.  Seems counter to the styles of every modern language with dynamic typing.

But also, I would just inline this.

File: service/src/main/resources/api/openapi.yml [hunk 8]
```diff
@@ -35,9 +35,56 @@ paths:
...
+      operationId: userProfileSet
+      tags: [ Profile ]
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/AnyObject'
+      responses:
+        '204':
```
Comment (cbookg) [inline service/src/main/resources/api/openapi.yml]: What if we just return the object is the "put" method as well (in the same format as "get")?

Seems like a pretty standard thing to do.  Don't have a great use case, but the caller may want to use the result object in the same way as a read, without building it themselves?

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    profileService.setProperty(getUser(), parsePath(path), body.getValue());
+
+    return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+  }
+
+  @Override
+  public ResponseEntity<AnyObject> userProfileLookup(String path) {
```
Comment (cbookg) [inline service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: why not userProfileGet?

File: service/src/main/java/bio/terra/user/db/ProfileDao.java (hunk 4 above)
Comment (cbookg) [reply service/src/main/java/bio/terra/user/db/ProfileDao.java]: +1, not an empty string :)

File: service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java (hunk 5 above)
Comment (cbookg) [reply service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java]: +1, no need for random.  normally would only do this on some kind of integration test when you need to avoid conflicts with other tests

File: service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+    mockMvc
+        .perform(put(API).contentType(MediaType.APPLICATION_JSON).content(json2))
+        .andExpect(status().isNoContent());
+
+    mockMvc
+        .perform(get(API))
+        .andExpect(status().isOk())
+        .andExpect(content().contentType(MediaType.APPLICATION_JSON))
+        .andExpect(jsonPath("$.value").value(""));
```
Comment (cbookg) [inline service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java]: surprised this is an empty string?

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java (hunk 9 above)
Comment (melissachang) [reply service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: We usually have verb first. getFolder, getPao, etc

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    this.profileService = profileService;
+    this.userFactory = userFactory;
+  }
+
+  private SamUser getUser() {
+    return userFactory.from(servletRequest);
+  }
+
+  private List<String> parsePath(String path) {
+    if (path == null || path.length() == 0) return List.of();
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: nit: StringUtils.isEmpty

File: service/src/main/java/bio/terra/user/app/controller/ProfileSamUserFactory.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+import bio.terra.common.iam.SamUser;
+import bio.terra.common.iam.SamUserFactory;
+import bio.terra.user.app.configuration.SamConfiguration;
+import javax.servlet.http.HttpServletRequest;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ProfileSamUserFactory {
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/app/controller/ProfileSamUserFactory.java:11]: `SamUserFactory`? I don't understand what's the profile prefix stand for

File: service/src/main/java/bio/terra/user/db/ProfileDao.java [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @ReadTransaction
+  public String getProperty(String userId, List<String> path) {
+    return getPropertySingle(userId, path);
+  }
+
+  private String getPropertySingle(String userId, List<String> path) {
+    final String sql =
+        "SELECT profile_obj #> :path::text[] AS value "
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/db/ProfileDao.java]: use
"""
"""

same for the rest

File: service/src/main/java/bio/terra/user/db/ProfileDao.java [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+@Component
+public class ProfileDao {
+  private final NamedParameterJdbcTemplate jdbcTemplate;
+
+  @Autowired
+  public ProfileDao(NamedParameterJdbcTemplate jdbcTemplate) {
+    this.jdbcTemplate = jdbcTemplate;
+  }
+
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/db/ProfileDao.java:22]: can you add javadoc? especially the logic from line 26 to 32

From what it does look like setProperties to me

File: service/src/main/java/bio/terra/user/db/ProfileDao.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+  @WriteTransaction
+  public void setProperty(String userId, List<String> path, String value) {
+    createRowIfNotExists(userId);
+
+    for (int i = 0; i < path.size() - 1; i++) {
+      var subPath = path.subList(0, i + 1);
+
+      if (!pathExists(userId, subPath)) {
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/db/ProfileDao.java]: this is a lot of sql query.. is there a way to merge some?

File: service/src/main/java/bio/terra/user/db/ProfileDao.java [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+            return rs.getString("value");
+          });
+    } catch (EmptyResultDataAccessException e) {
+      throw new PropertyNotFoundException("", e);
+    }
+  }
+
+  private void createRowIfNotExists(String userId) {
+    final String sql =
+        "INSERT INTO user_profile (user_id, profile_obj) "
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/db/ProfileDao.java]: you can just set a defaultValue{} to project_obj

File: service/src/main/java/bio/terra/user/service/user/ProfileService.java [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  public void setProperty(SamUser user, List<String> path, Object value) {
+    var userId = user.getSubjectId();
+    var json = toJson(value);
+
+    profileDao.setProperty(userId, path, json);
+  }
+
+  private String toJson(Object value) {
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/service/user/ProfileService.java]: i think this is an over-refactoring. you didn't do this for readValue either.

File: service/src/main/resources/db/changesets/20221012_initial.yaml [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+                  type: text
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+                  remarks: Sam user ID
+              - column:
+                  name: profile_obj
+                  type: jsonb
+                  constraints:
+                    nullable: false
```
Comment (yuhuyoyo) [inline service/src/main/resources/db/changesets/20221012_initial.yaml:20]: defaultValue: "{}"

File: service/src/main/java/bio/terra/user/db/ProfileDao.java [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+import bio.terra.common.db.WriteTransaction;
+import bio.terra.user.service.exception.PropertyNotFoundException;
+import java.util.List;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Component;
+
+@Component
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/db/ProfileDao.java:14]: please add unit test for Dao

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+
+    return Arrays.asList(path.split("\\."));
+  }
+
+  @Override
+  public ResponseEntity<Void> userProfileSet(AnyObject body, String path) {
+    profileService.setProperty(getUser(), parsePath(path), body.getValue());
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java:35]: should we throw 401  if parsePath returns empty array?

File: service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java (hunk 5 above)
Comment (iliadsh) [reply service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java]: @melissachang Since it's using a real db I wasn't sure if there would be conflicts between tests. Should I put a cleanup step in the beforeEach instead?

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java (hunk 20 above)
Comment (iliadsh) [reply service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java:35]: Agree, since destroying the root seems a bit dangerous. The way the query is set up makes it a no-op if you try that anyway.

File: service/src/main/java/bio/terra/user/db/ProfileDao.java (hunk 16 above)
Comment (iliadsh) [reply service/src/main/java/bio/terra/user/db/ProfileDao.java]: Hm, I'm not sure I understand. I believe the default value is {}, but how would it work when the row for the user doesn't exist yet?

Comment (iliadsh) [reply service/src/main/java/bio/terra/user/db/ProfileDao.java]: Oh wait, do you mean keeping this query but just not inserting '{}' along with the id?

Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/user/db/ProfileDao.java]: i guess I don't understand why you need this step. if you add defaultValue to that column, when you add a row for a new user, even if you dont' have any profile_obj to set, the database will set a {} for you.

File: service/src/main/java/bio/terra/user/app/controller/ProfileSamUserFactory.java (hunk 12 above)
Comment (iliadsh) [reply service/src/main/java/bio/terra/user/app/controller/ProfileSamUserFactory.java:11]: SamUserFactory also exists in the terra common lib, and is being used in this class too, so it avoids name collisions

File: service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java (hunk 5 above)
Comment (melissachang) [reply service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java]: In general, it's good practice to make your test resilient. Other tests could be running at the same time, other tests could have or have not run before or after this.

For example for WSM local development, I'll reuse a workspace between runs (as opposed to creating a new workspace for each test, which is what normally happens.) That workspace will have leftover resources from previous runs.

You don't need to clean up db. Instead, I think it's simpler to do something like this:

```
  @Test
  void recursiveSet() throws Exception {
    String path = TestUtils.appendRandomNumber("user.name.first");
    setUserProfile(path, "John");
    assertUserProfile(path, "John");
  }
```
One advantage of not cleaning up db is that if a test fails, you can inspect db.

File: service/src/main/resources/db/changesets/20221013_pathfunc.yaml [hunk 21]
```diff
@@ -0,0 +1,8 @@
+databaseChangeLog:
+  - changeSet:
```
Comment (yuhuyoyo) [inline service/src/main/resources/db/changesets/20221013_pathfunc.yaml]: merge those two files into one cuz apprently liquibase merging is pretty time costly.

File: service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java (hunk 5 above)
Comment (iliadsh) [reply service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java]: I think in that case it would be best to separate the tests by using random users and thus entire profile objects rather than randomizing the paths, because if several tests use the same parent nodes like `user` then they could interfere even if the leaves like `.name.first` are randomized.

Comment (melissachang) [reply service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java]: Agree, that sounds better

File: service/src/main/java/bio/terra/user/db/ProfileDao.java (hunk 19 above)
Comment (cbookg) [reply service/src/main/java/bio/terra/user/db/ProfileDao.java:14]: Could you be more specific?  I wouldn't want to just duplicate the higher level tests that are exercising all the db functionality?  Or is there something to be tested at this level that isn't tested above?

Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/user/db/ProfileDao.java:14]: the controller test shouldn't be testing the implementation details of a db test. it's generally good practice to have unit test per class so you don't cram all the edge cases test cases to a higher level class that can be calling multiple services/db. 

the controller test should be testing its interaction with the profile service.

Comment (melissachang) [reply service/src/main/java/bio/terra/user/db/ProfileDao.java:14]: Note - I disagree with this good practice. :) I like black box testing that does what a user does. The controller tests make API calls like clients would.

Dao could have tests as well. My preference is to have controller tests for most functionality, especially basic functionality.

Comment (cbookg) [reply service/src/main/java/bio/terra/user/db/ProfileDao.java:14]: I disagree with testing the same things in both places.

Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/user/db/ProfileDao.java:14]: based on the definition of unit test https://en.wikipedia.org/wiki/Unit_testing, i stand that we should strive for having unit level tests per classes. 

i don't understand why this is contentious. Seems a general practice in databiosphere/ or google3.

Comment (cbookg) [reply service/src/main/java/bio/terra/user/db/ProfileDao.java:14]: Took another look, and I think it's a fair point that some of the current tests can be moved down to the Dao.  The service doesn't care about the tree structure details.  So some tests from the API layer including success, top level errors, multiple users.  And then dao specific tests around data structure.

Comment (melissachang) [reply service/src/main/java/bio/terra/user/db/ProfileDao.java:14]: Sorry I realized I wasn't clear. No need to change anything, just wanted to clarify. I wasn't saying we shouldn't have unit tests. Here's what we did on past teams:

1. Keep all original controller tests as controller tests. They provide good coverage of API.
2. In ProfileDaoTest, have a unit test for each dao method. And possibly corner case tests (setProperty_rowExists(), setProperty_rowDoesNotExist())

You have dao unit tests to make sure dao works. And you have API tests for the various ways API can be called.

**Review by cbookg — APPROVED** [review]

File: service/src/main/java/bio/terra/user/db/ProfileDao.java [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+   *
+   * <p>b) If a path element exists but is not an indexible type (object or array), then the
+   * operation will abort.
+   *
+   * @param userId - Sam user ID
+   * @param path - path to potentially nested property
+   * @param value - JSON serialized value
+   */
+  @WriteTransaction
+  public void setProperty(String userId, List<String> path, String value) {
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/db/ProfileDao.java:38]: can we pass in json value here?

File: service/src/main/java/bio/terra/user/service/user/ProfileService.java [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+  public void setProperty(SamUser user, List<String> path, Object value) {
+    if (path.size() == 0) {
+      throw new InvalidPropertyException("Cannot overwrite the root object.");
+    }
+
+    var userId = user.getSubjectId();
+    try {
+      var json = objectMapper.writeValueAsString(value);
+
+      profileDao.setProperty(userId, path, json);
```
Comment (melissachang) [inline service/src/main/java/bio/terra/user/service/user/ProfileService.java]: nit can this be moved out of try/catch block?

File: service/src/main/java/bio/terra/user/service/user/ProfileService.java [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+    } catch (JsonProcessingException e) {
+      throw new MalformedPropertyException("Value to set must be valid JSON", e);
```
Comment (melissachang) [inline service/src/main/java/bio/terra/user/service/user/ProfileService.java]: Useful to print invalid value

throw new MalformedPropertyException(String.format("Value to set must be valid JSON. Value: %s", value), e);

File: service/src/main/java/bio/terra/user/service/user/ProfileService.java [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+  public Object getProperty(SamUser user, List<String> path) {
+    var userId = user.getSubjectId();
+
+    var prop = profileDao.getProperty(userId, path);
+    if (prop == null) return null;
+
+    try {
+      return objectMapper.readValue(prop, JsonNode.class);
+    } catch (JsonProcessingException e) {
+      throw new InternalServerErrorException("Failed to process property json", e);
```
Comment (melissachang) [inline service/src/main/java/bio/terra/user/service/user/ProfileService.java]: Print prop in error message

File: service/src/main/resources/api/openapi.yml [hunk 26]
```diff
@@ -35,9 +35,52 @@ paths:
               schema:
                 $ref: '#/components/schemas/VersionProperties'
 
+  # User profile management
+  /api/profile:
+    parameters:
+      - $ref: '#/components/parameters/ProfilePath'
+    get:
+      summary: Retrieve a profile property.
```
Comment (melissachang) [inline service/src/main/resources/api/openapi.yml]: Nice to add details (summary or description), for example for someone calling this API from swagger. If ProfilePath is `/`, will they get entire user profile?

For setUserProfile, ProfilePath must have at least one "directory".

Also good to mention that entire user profile is returned, not just the part that was set. (Should there be a test for this?)

Also, setUserProfile is missing `parameters`, not sure how things are working..

File: service/src/main/resources/db/changesets/20221012_initial.yaml [hunk 27]
```diff
@@ -0,0 +1,25 @@
...
+                  defaultValue: "{}"
+                  remarks: Schemaless object storing the profile data
+        -  sqlFile:
+            splitStatements:  false
+            path:  db/functions/pathfunc.sql
```
Comment (melissachang) [inline service/src/main/resources/db/changesets/20221012_initial.yaml]: nit can you have `relativeToChangelogFile`, so `functions` is in `changesets` directory. (Don't need `db` dir.)

https://cs.github.com/broadinstitute/cromwell/blob/1a3889689060dcff2b6706a98ae69fd2fa9f70ef/database/migration/src/main/resources/changesets/failure_metadata.xml?q=path%3Achangeset+%22sqlFile%22#L17

File: service/src/main/resources/db/functions/pathfunc.sql [hunk 28]
```diff
@@ -0,0 +1,42 @@
+-- pathRecurse(target, path, new_value) is a wrapper over jsonb_set which essentially performs:
```
Comment (melissachang) [inline service/src/main/resources/db/functions/pathfunc.sql]: Thank you for these comments!

File: service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java (hunk 7 above)
Comment (melissachang) [reply service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java]: I don't see this test (I might have missed it)

**Review by melissachang — APPROVED** [review]

**Review by yuhuyoyo — APPROVED** [review]

File: service/src/main/resources/api/openapi.yml (hunk 26 above)
Comment (iliadsh) [reply service/src/main/resources/api/openapi.yml]: Yep fair enough, those descriptions could use some better documentation.

As for the others, ATM the put endpoint should be returning the value that was just passed in by calling getUserProfile on the same path. Parameters - they are specified at the top level so both the get and put should use them is my understanding?

Comment (melissachang) [reply service/src/main/resources/api/openapi.yml]: SG, I misread the code, thanks

File: service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java (hunk 7 above)
Comment (iliadsh) [reply service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java]: Ah it was removed since we agreed that setting the root directly should not be allowed.

`clobberSet` in the dao tests is probably the closest one since it also tests overwrite behavior, but it doesn't have an interesting tree structure, so maybe there should be a new test that is closer to this.
